### ISSUE TITLE
[feature] different approach for testing complicated rules

### ISSUE THREAD

**Issue by mrwsl** [ast-grep/ast-grep#1555]
# ⭐ Suggestion

The current testing approach is good for rules that match one line. If your rules need more context, the current tests can be pretty large with a lot of code repetition. [Here is my example](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImNwcCIsInF1ZXJ5IjoiY29uc29sZS5sb2coJE1BVENIKSIsInJld3JpdGUiOiJsb2dnZXIubG9nKCRNQVRDSCkiLCJzdHJpY3RuZXNzIjoic21hcnQiLCJzZWxlY3RvciI6IiIsImNvbmZpZyI6InJ1bGU6XG4gIGtpbmQ6IGZ1bmN0aW9uX2RlY2xhcmF0b3JcbiAgaGFzOlxuICAgIGtpbmQ6IG9wZXJhdG9yX25hbWVcbiAgICByZWdleDogJ1xcK3wtfFxcKnwvfCV8PT18IT18PHw+fDw9fD49fCYmfFxcfFxcfHwmfFxcfHxcXF58PDx8Pj58Ojp8PT58ISdcbiAgaW5zaWRlOlxuICAgIGtpbmQ6IGZ1bmN0aW9uX2RlZmluaXRpb25cbiAgICBpbnNpZGU6XG4gICAgICBraW5kOiBmaWVsZF9kZWNsYXJhdGlvbl9saXN0XG4gIG5vdDpcbiAgICBoYXM6XG4gICAgICBvcGVyYXRvcl9uYW1lOlxuICAgICAgcmVnZXg6ICdcXFtcXF18ZGVsZXRlfFxcLCciLCJzb3VyY2UiOiIjaW5jbHVkZSA8aW9zdHJlYW0+XG5cbmNsYXNzIE15TnVtYmVyIHtcbnB1YmxpYzpcbiAgICBpbnQgdmFsdWU7XG5cbiAgICBNeU51bWJlcihpbnQgdikgOiB2YWx1ZSh2KSB7fVxuXG4gICAgY29uc3RleHByIFZlYzwgVCA+JiBvcGVyYXRvcltdKCBjb25zdCBzaXplX3QgbGF5ZXIgKSBub2V4Y2VwdCB7IHJldHVybiBhY2Nlc3M8IHRydWUgPiggbGF5ZXIgKTsgfVxuXG4gICAgY29uc3RleHByIFQmIG9wZXJhdG9yW10oIGNvbnN0IHNpemVfdCBsYXllciwgY29uc3Qgc2l6ZV90IHBhdGggKSBub2V4Y2VwdCB7IHJldHVybiBhY2Nlc3M8IHRydWUgPiggbGF5ZXIsIHBhdGggKTsgfVxuXG4gICAgY29uc3RleHByIGRlY2x0eXBlKGF1dG8pICBvcGVyYXRvcltdKCBjb25zdCBzaXplX3QgbGF5ZXIsIGNvbnN0IHNpemVfdCBwYXRoLCBjb25zdCBzaXplX3QgcG9pbnQgKSBub2V4Y2VwdCB7IHJldHVybiBhY2Nlc3M8IHRydWUgPiggbGF5ZXIsIHBhdGgsIHBvaW50ICk7IH1cblxuICAgIG9wZXJhdG9yIGJvb2woKSBjb25zdCB7IHJldHVybiB2YWx1ZSAhPSAwOyB9XG5cbiAgICBkZWNsdHlwZShhdXRvKSBnZXRUeXBlKCkgY29uc3QgeyByZXR1cm4gdmFsdWU7IH0gICAgIC8vIE9wZXJhdG9yIGRlY2x0eXBlXG5cbiAgICBNeU51bWJlciogb3BlcmF0b3ItPigpIHsgcmV0dXJuIHRoaXM7IH0gICAgIC8vIE9wZXJhdG9yIC0+XG5cblxuICAgIE15TnVtYmVyIG9wZXJhdG9yLChjb25zdCBNeU51bWJlciYgb3RoZXIpIGNvbnN0IHsgcmV0dXJuIG90aGVyOyB9ICAgICAvLyBPcGVyYXRvciAsXG5cblxuICAgIHZvaWQgb3BlcmF0b3IgQCAoY29uc3QgTXlOdW1iZXIgb3RoZXIpIGNvbnN0IHsgLyogTm8tb3AsIGp1c3QgZm9yIHRlc3RpbmcgKi8gfSAgICAgLy8gT3BlcmF0b3IgQFxuXG5cbiAgICB2b2lkKiBvcGVyYXRvciBuZXcoc2l6ZV90IHNpemUpIHsgcmV0dXJuIDo6b3BlcmF0b3IgbmV3KHNpemUpOyB9ICAgICAvLyBPcGVyYXRvciBuZXcgYW5kIGRlbGV0ZVxuXG4gICAgdm9pZCBvcGVyYXRvciBkZWxldGUodm9pZCogcHRyKSB7IDo6b3BlcmF0b3IgZGVsZXRlKHB0cik7IH1cblxuXG4gICAgaW50JiBvcGVyYXRvcltdKHNpemVfdCBpbmRleCkgeyAgICAgLy8gT3BlcmF0b3IgW11cbiAgICBzdGF0aWMgaW50IGR1bW15ID0gMDsgIFxuICAgICAgICBpZiAoaW5kZXggPT0gMCkgeyByZXR1cm4gdmFsdWU7XG4gICAgICAgIH0gZWxzZSB7IHJldHVybiBkdW1teTsgfVxuICAgIH1cblxuICAgIC8vIFZpb2xhdGVzIHRoZSBndWlkZWxpbmU6IG9wZXJhdG9ycyBhcmUgaW1wbGVtZW50ZWQgYXMgbWVtYmVyIGZ1bmN0aW9uc1xuICAgIE15TnVtYmVyIG9wZXJhdG9yKyhjb25zdCBNeU51bWJlciYgb3RoZXIpIGNvbnN0IHsgcmV0dXJuIE15TnVtYmVyKHRoaXMtPnZhbHVlICsgb3RoZXIudmFsdWUpOyB9XG5cbiAgICBNeU51bWJlciBvcGVyYXRvci0oY29uc3QgTXlOdW1iZXImIG90aGVyKSBjb25zdCB7IHJldHVybiBNeU51bWJlcih0aGlzLT52YWx1ZSAtIG90aGVyLnZhbHVlKTsgfVxuXG4gICAgTXlOdW1iZXIgb3BlcmF0b3IqKGNvbnN0IE15TnVtYmVyJiBvdGhlcikgY29uc3QgeyByZXR1cm4gTXlOdW1iZXIodGhpcy0+dmFsdWUgKiBvdGhlci52YWx1ZSk7IH1cblxuICAgIE15TnVtYmVyIG9wZXJhdG9yLyhjb25zdCBNeU51bWJlciYgb3RoZXIpIGNvbnN0IHsgcmV0dXJuIE15TnVtYmVyKHRoaXMtPnZhbHVlIC8gb3RoZXIudmFsdWUpOyB9XG5cbiAgICBNeU51bWJlciBvcGVyYXRvciUoY29uc3QgTXlOdW1iZXImIG90aGVyKSBjb25zdCB7IHJldHVybiBNeU51bWJlcih0aGlzLT52YWx1ZSAlIG90aGVyLnZhbHVlKTsgfVxuXG4gICAgYm9vbCBvcGVyYXRvcj09KGNvbnN0IE15TnVtYmVyJiBvdGhlcikgY29uc3QgeyByZXR1cm4gdGhpcy0+dmFsdWUgPT0gb3RoZXIudmFsdWU7IH1cblxuICAgIGJvb2wgb3BlcmF0b3IhPShjb25zdCBNeU51bWJlciYgb3RoZXIpIGNvbnN0IHsgcmV0dXJuIHRoaXMtPnZhbHVlICE9IG90aGVyLnZhbHVlOyB9XG5cbiAgICBib29sIG9wZXJhdG9yPChjb25zdCBNeU51bWJlciYgb3RoZXIpIGNvbnN0IHsgcmV0dXJuIHRoaXMtPnZhbHVlIDwgb3RoZXIudmFsdWU7IH1cblxuICAgIGJvb2wgb3BlcmF0b3I8PShjb25zdCBNeU51bWJlciYgb3RoZXIpIGNvbnN0IHsgcmV0dXJuIHRoaXMtPnZhbHVlIDw9IG90aGVyLnZhbHVlOyB9XG5cbiAgICBib29sIG9wZXJhdG9yPihjb25zdCBNeU51bWJlciYgb3RoZXIpIGNvbnN0IHsgcmV0dXJuIHRoaXMtPnZhbHVlID4gb3RoZXIudmFsdWU7IH1cblxuICAgIGJvb2wgb3BlcmF0b3I+PShjb25zdCBNeU51bWJlciYgb3RoZXIpIGNvbnN0IHsgcmV0dXJuIHRoaXMtPnZhbHVlID49IG90aGVyLnZhbHVlOyB9XG5cbiAgICBNeU51bWJlciBvcGVyYXRvciYoY29uc3QgTXlOdW1iZXImIG90aGVyKSBjb25zdCB7IHJldHVybiBNeU51bWJlcih0aGlzLT52YWx1ZSAmIG90aGVyLnZhbHVlKTsgfVxuXG4gICAgTXlOdW1iZXIgb3BlcmF0b3J8KGNvbnN0IE15TnVtYmVyJiBvdGhlcikgY29uc3QgeyByZXR1cm4gTXlOdW1iZXIodGhpcy0+dmFsdWUgfCBvdGhlci52YWx1ZSk7IH1cblxuICAgIE15TnVtYmVyIG9wZXJhdG9yXihjb25zdCBNeU51bWJlciYgb3RoZXIpIGNvbnN0IHsgcmV0dXJuIE15TnVtYmVyKHRoaXMtPnZhbHVlIF4gb3RoZXIudmFsdWUpOyB9XG5cbiAgICBNeU51bWJlciBvcGVyYXRvcjw8KGNvbnN0IE15TnVtYmVyJiBvdGhlcikgY29uc3QgeyByZXR1cm4gTXlOdW1iZXIodGhpcy0+dmFsdWUgPDwgb3RoZXIudmFsdWUpOyB9XG5cbiAgICBNeU51bWJlciBvcGVyYXRvcj4+KGNvbnN0IE15TnVtYmVyJiBvdGhlcikgY29uc3QgeyByZXR1cm4gTXlOdW1iZXIodGhpcy0+dmFsdWUgPj4gb3RoZXIudmFsdWUpOyB9XG59O1xuIn0=) where some context (the class in this example but we have way more complex examples) is needed. In this case, you need to have the same code over and over again to test each operator from rule's line 5 for proper testing. This leads quickly to large test files. I am very happy that I discovered ast-grep - coming from semgrep. One thing semgrep did good in my opinion was the [testing](https://semgrep.dev/docs/writing-rules/testing-rules#example): use comments to indicate matches/non-matches inside your example code. 

# 💻 Use Cases

Currently, we are copying our large code examples for testing into parts that only have one match. This leads to large test files.

**Comment by HerringtonDarkholme** [comment]
This will be a breaking change. Or we will introduce a new testing format.

We need more feedback to adopt it.

**Comment by HerringtonDarkholme** [comment]
With https://github.com/ast-grep/ast-grep/pull/1561, you can override the `unused-suppression` rule's severity.

Combining this feature, users can write rule testing in one single file. For example, user can write several lines of valid code without ignore comments and invalid code with ignore comments.

If a rule is implemented wrong, either an unexpected error will be reported, or an unused suppression is reported in the command `sg scan --error=unused-suppression`


```ts
// ast-grep-ignore
should_report_error()
should_not_report_error()
```

for example, `sg scan --error=unused-suppression` will not report error for the file above, unless the rule itself is implemented wrong.

**Comment by mrwsl** [comment]
That is a really nice and helpful addition! Just to make sure I understand correctly: the PR does not provide a functionality to tell ast-grep that it should expect invalid code at certain lines, correct? Background is for example that I want to test multiple occurences of invalid code and ast-grep should know how many of them there are.

**Comment by HerringtonDarkholme** [comment]
> tell ast-grep that it should expect invalid code at certain lines, 

You can do that by using the magic comment `ast-grep-ignore`.

>  I want to test multiple occurences of invalid code 

You can have one test file that contains multiple errors and suppressions. For example

```
#include <iostream>

class MyNumber {
public:
    int value;

    MyNumber(int v) : value(v) {} // this line is good, no suppression

    // Violates the guideline: operators are implemented as member functions
    MyNumber operator+(const MyNumber& other) const { return MyNumber(this->value + other.value); } // ast-grep-ignore

    MyNumber operator-(const MyNumber& other) const { return MyNumber(this->value - other.value); } } // ast-grep-ignore

    MyNumber operator*(const MyNumber& other) const { return MyNumber(this->value * other.value); } } // ast-grep-ignore

    MyNumber operator/(const MyNumber& other) const { return MyNumber(this->value / other.value); } } // ast-grep-ignore
 

};

```

**Comment by HerringtonDarkholme** [comment]
released in 0.29.0
